pkg/controller/infrastructure/terraform_reconciler.go

```diff
@@ -0,0 +1,255 @@
+	return &TerraformReconciler{
+		client:                     client,
+		restConfig:                 restConfig,
+		log:                        log,
+		disableProjectedTokenMount: disableProjectedTokenMount,
+	}
+}
+
+// Restore restores the infrastructure after a control plane migration. Effectively it performs a recovery of data from the infrastructure.status.state and
+// proceeds to reconcile.
+func (t *TerraformReconciler) Restore(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster) error {
+	var initializer terraformer.StateConfigMapInitializer
+
+	terraformState, err := terraformer.UnmarshalRawState(infra.Status.State)
+	if err != nil {
+		return err
+	}
+
+	initializer = terraformer.CreateOrUpdateState{State: &terraformState.Data}
+	return t.reconcile(ctx, infra, cluster, initializer)
+}
+
+// Reconcile reconciles infrastructure using Terraformer.
+func (t *TerraformReconciler) Reconcile(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster) error {
+	err := t.reconcile(ctx, infra, cluster, terraformer.StateConfigMapInitializerFunc(terraformer.CreateState))
+	return util.DetermineError(err, helper.KnownCodes)
+}
+
+func (t *TerraformReconciler) reconcile(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster, initializer terraformer.StateConfigMapInitializer) error {
+	log := t.log
+
+	log.Info("reconcile infrastructure using terraform reconciler")
+	config, err := helper.InfrastructureConfigFromInfrastructure(infra)
+	if err != nil {
+		return err
+	}
+
+	terraformFiles, err := infrastructure.RenderTerraformerTemplate(infra, config, cluster)
+	if err != nil {
+		return err
+	}
+
+	// need to know if application credentials are used
+	credentials, err := openstack.GetCredentials(ctx, t.client, infra.Spec.SecretRef, false)
+	if err != nil {
+		return err
+	}
+
+	tf, err := internal.NewTerraformerWithAuth(log, t.restConfig, infrastructure.TerraformerPurpose, infra, credentials, t.disableProjectedTokenMount)
+	if err != nil {
+		return util.DetermineError(err, helper.KnownCodes)
+	}
+
+	if err := tf.
+		InitializeWith(ctx, terraformer.DefaultInitializer(t.client, terraformFiles.Main, terraformFiles.Variables, terraformFiles.TFVars, initializer)).
+		Apply(ctx); err != nil {
+		return fmt.Errorf("failed to apply the terraform config: %w", err)
+	}
+
+	status, state, err := t.computeTerraformStatusState(ctx, tf, config)
+	if err != nil {
+		return err
+	}
+
+	return infrastructure.PatchProviderStatusAndState(ctx, t.client, infra, status, state)
+}
+
+// Delete deletes the infrastructure using Terraformer.
+func (t *TerraformReconciler) Delete(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, c *extensions.Cluster) error {
+	return util.DetermineError(t.delete(ctx, infra, c), helper.KnownCodes)
```

Comment:
Some codepaths in `delete` already call `DetermineError`. If this needs to be done always, we might _just_ do it here. Otherwise, if this does not need to be done in _some_ cases I would remove this call to `DetermineError`